internal/middleware/middleware.go

```diff
@@ -0,0 +1,148 @@
+	status   int
+	response string
+	http.ResponseWriter
+}
+
+func (w *respWriterWithStatus) WriteHeader(code int) {
+	w.status = code
+	w.ResponseWriter.WriteHeader(code)
+}
+
+func (w *respWriterWithStatus) Write(d []byte) (int, error) {
+	w.response = string(d)
+	return w.ResponseWriter.Write(d)
+}
+
+func NewTransactionMgmtMiddleware(cfg *svcCfg.SvcConfig) *TransactionMgmtMiddleware {
+	return &TransactionMgmtMiddleware{
+		cfg:    cfg.Cfg,
+		jwt:    cfg.JwtSvc.JwtSvc,
+		cacher: cfg.Cacher.Cacher,
+	}
+}
+func (u TransactionMgmtMiddleware) ExtractUser(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		cookie, err := r.Cookie("token")
+		if err != nil {
+			log.Error(err)
+			response.ToJson(w, http.StatusUnauthorized, codes.GetErr(codes.ErrUnauthorized), nil)
+			return
+		}
+		if cookie.Value == "" {
+			log.Error(err)
+			response.ToJson(w, http.StatusUnauthorized, codes.GetErr(codes.ErrUnauthorized), nil)
+			return
+		}
+		token, err := u.jwt.ValidateToken(cookie.Value)
+		if err != nil {
+			log.Error(err)
+			if strings.Contains(err.Error(), "Token is expired") {
+				response.ToJson(w, http.StatusUnauthorized, codes.GetErr(codes.ErrTokenExpired), nil)
+				return
+			}
+			response.ToJson(w, http.StatusUnauthorized, codes.GetErr(codes.ErrMatchingToken), nil)
+			return
+		}
+		if !token.Valid {
+			response.ToJson(w, http.StatusUnauthorized, codes.GetErr(codes.ErrUnauthorized), nil)
+			return
+		}
+		mapClaims, ok := token.Claims.(jwt.MapClaims)
+		if !ok {
+			response.ToJson(w, http.StatusInternalServerError, codes.GetErr(codes.ErrAssertClaims), nil)
+			return
+		}
+		userId, ok := mapClaims["user_id"]
+		if !ok {
+			response.ToJson(w, http.StatusBadRequest, codes.GetErr(codes.ErrAssertUserid), nil)
+			return
+		}
+		userIdStr, ok := userId.(string)
+		if !ok {
+			response.ToJson(w, http.StatusBadRequest, codes.GetErr(codes.ErrAssertUserid), nil)
+			return
+		}
+		sessionStruct := model.SessionStruct{UserId: userIdStr, Cookie: cookie.Value}
+		ctx := session.SetSession(r.Context(), sessionStruct)
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+func (t TransactionMgmtMiddleware) Cacher(requireAuth bool) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			var key string
+			var cacheResponse model.CacheResponse
+			key = fmt.Sprint(r.URL.String())
+			if requireAuth != false {
+				sessionStruct := session.GetSession(r.Context())
+				session, ok := sessionStruct.(model.SessionStruct)
+				if !ok {
+					response.ToJson(w, http.StatusBadRequest, codes.GetErr(codes.ErrAssertUserid), nil)
+					return
+				}
+				key = fmt.Sprint(key + "/auth/" + session.UserId)
+			}
+			Cacher := t.cacher
+			by, err := Cacher.Get(key)
+			if err == nil {
+				err = json.Unmarshal(by, &cacheResponse)
+				if err != nil {
+					log.Error(err)
+					return
```

Comment:
Appropriate response needs to be returned to the user without revealing the underlying usage of redis/cache here.
Something like: failed to fetch response with 500 error.
Test case also needs update.

Reply:
done

Reply:
Where is this implemented?

Reply:
Where is this implemented?

Reply:
done

Reply:
Why is `ErrGetTransaction` error code returned? The error code should specify a generic error while processing the request.

Reply:
done

Reply:
> Appropriate response needs to be returned to the user without revealing the underlying usage of redis/cache here. Something like: failed to fetch response with 500 error. Test case also needs update.

done